### PR TITLE
@alloy => [Toolchain] Add Babel 7

### PR DESCRIPTION
**Pull request by damassi**
This PR upgrades Force with Babel 7 for TypeScript support and more efficiently interfaces with Reaction.  New development flow:

```bash
cd artsy/reaction
yarn link
yarn watch // concurrently boot babel -w and babel-relay -w

cd force 
yarn link @artsy/reaction
yarn start 
```

Changes made to reaction will automatically be reloaded in force. 

(In a future PR the above flow will be simplified behind a single command when in development.)

Additionally, `import Foo from '@artsy/reaction/dist/Foo'` has been aliased to `import Foo from 'reaction/Foo'`

**NOTE:** `babel-plugin-rewire` has been removed as it is incompatable, and instead we're using the regular `rewire` package. Example usecase:

```javascript 
// Component to test 
import _SomeComponentToMock from '/path/to/component'`

let SomeComponentToMock = _SomeComponentToMock

export default SomeClass extends Component {
 ...
}

// test.test.js
const rewire = require('rewire')('../path/to/component') 
const SomeClass = rewire.default

describe('test', () => {
  it('works', () => {
    rewire.__set__('SomeComponentToMock', () => <div />)
  })
})
```

#### TODO: 
- [x] Make notice about new `reaction/Components` alias to `@artsy/reaction/dist/Components`

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,33 +1,32 @@
...
       "plugins": [
         ["babel-plugin-styled-components", {
-          "ssr": true,
-          "displayName": true
-        }]
-      ]
-    },
-    "test" : {
-      "plugins": [
-        "babel-plugin-rewire"
```
Comment (alloy) [inline .babelrc]: 👋

File: .babelrc [hunk 2]
```diff
@@ -1,33 +1,32 @@
...
-    "stage-3"
+    "@babel/react",
+    "@babel/stage-3",
+    "@babel/preset-typescript"
   ],
   "plugins": [
+    "@babel/plugin-transform-runtime",
+    "@babel/plugin-proposal-class-properties",
+    "@babel/plugin-transform-typescript",
+    ["relay", { "artifactDirectory": "/node_modules/@artsy/reaction-force/src/__generated__" }],
```
Comment (alloy) [inline .babelrc]: Did you _need_ to make this an absolute path?

File: test.config.js [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
 require('coffeescript/register')
-require('source-map-support/register')
-require('babel-polyfill')
+require('@babel/polyfill')
 require('raf/polyfill')
 require('should')
 require('./src/lib/jade_hook')
 
+// FIXME: Do we need this?
+// require('source-map-support/register')
```
Comment (alloy) [inline test.config.js]: Once we do AOT compilation we probably want to re-enable this on the server in development mode only.

**Comment by alloy** [comment]
Note to self as well as @damassi: Don’t forget to cover starting the `relay-compiler` in watch mode, either separately or just by running `yarn start` inside Reaction (that really depends on how the overall DX is supposed to be going forward).

File: src/desktop/apps/article/components/App.js [hunk 4]
```diff
@@ -4,69 +4,65 @@ import ReactDOM from 'react-dom'
 import { Article } from '@artsy/reaction-force/dist/Components/Publishing'
 import ArticleLayout from './layouts/Article'
 import { EditButton } from 'desktop/apps/article/components/EditButton'
+import { hot } from 'react-hot-loader'
```
Comment (l2succes) [inline src/desktop/apps/article/components/App.js]: Why do we need to include `react-hot-loader` directly in our code now? 
Do we only need to wrap a top-level parent component or any component we want to enable hot reloading on?

Comment (damassi) [reply src/desktop/apps/article/components/App.js]: This is pretty annoying, I agree, but we were over-wrapping with the previous solution and it was incompatible in Babel 7. And yup, only top-level component; it cascades down the tree. I'm going to look into automating this somehow, but right now there aren't much options. 

(See https://github.com/gaearon/react-hot-loader/tree/next for more info)

**Comment by orta** [comment]
Tried setting this up locally, seeing 

```sh
raven@1.2.1 alert: uncaughtException: 1fd302955d7b499e8deda3152acc4f49
error: Forever detected script exited with code: 0
error: Script restart attempt #2
```

inside force, could be my env, will update comment

**Comment by damassi** [comment]
@orta - make sure `NODE_ENV=development` in `.env` file

**Comment by alloy** [comment]
When I took this for a spin last week everything worked well 👍 

There’s still an unticked TODO item in your original comment and I'd like to see some doc/README update on how people are supposed to start the dev environment.

**Comment by damassi** [comment]
@alloy - updated README, and will make a notice in #dev once merged about new paths

File: package.json [hunk 5]
```diff
@@ -47,24 +69,13 @@
     "artsy-morgan": "artsy/artsy-morgan",
     "artsy-xapp": "artsy/artsy-xapp",
     "async": "^2.1.2",
-    "babel": "^6.5.2",
-    "babel-core": "^6.17.0",
+    "babel-core": "^7.0.0-0",
     "babel-eslint": "^7.0.0",
-    "babel-loader": "^7.1.2",
+    "babel-loader": "8.0.0-beta.0",
```
Comment (orta) [inline package.json]: weird, you'd think they'd try keep this in sync

File: src/desktop/apps/article/components/__tests__/InfiniteScrollArticle.test.js [hunk 6]
```diff
@@ -41,13 +43,13 @@ describe('<InfiniteScrollArticle />', () => {
     window.history.replaceState.reset()
   })
 
-  it('renders the initial article', () => {
+  it.only('renders the initial article', () => {
```
Comment (orta) [inline src/desktop/apps/article/components/__tests__/InfiniteScrollArticle.test.js]: caution

**Comment by orta** [comment]
Alright, I've given this a second run through and everything seems to be going 👍 locally

Happy to merge this in ( congrats! ) and get this tested on staging.

**Comment by alloy** [comment]
Great stuff @damassi 👏

**Pull request merged**